samples/Mvc.ControllerForMultipleAreasSample/MultipleAreasControllerConvention.cs

```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ApplicationModels;
+
+namespace ControllerForMultipleAreasSample
+{
+    public class MultipleAreasControllerConvention : IApplicationModelConvention
+    {
+        public void Apply(ApplicationModel application)
+        {
+            var controllerModels = new List<ControllerModel>();
+            foreach (var controller in application.Controllers)
+            {
+                var areaNames = controller.ControllerType.GetCustomAttributes<MultipleAreasAttribute>().FirstOrDefault()?.AreaNames;
+                controller.RouteValues.Add("area", areaNames[0]);
+                for (var i = 1; i < areaNames?.Length; i++)
+                {
+                    var controllerCopy = new ControllerModel(controller);
+                    controllerCopy.RouteValues["area"] = areaNames[i];
+                    controllerModels.Add(controllerCopy);
+                }
+            }
+
+            foreach (var model in controllerModels)
+            {
+                application.Controllers.Add(model);
```

Comment:
Any reason not to do this within the inner nested loop above? That is, instead of making a list, then iterating over the list, why not just add each new model immediately?

Reply:
Because that modifies the collection (application.Controllers), which halts the enumeration (in the above for loop).

Reply:
Oh, obviously, thanks! 😄